src/trio/testing/_exceptiongroup_util.py

```diff
@@ -0,0 +1,385 @@
+
+E = TypeVar("E", bound=BaseException)
+E2 = TypeVar("E2", bound=BaseException)
+T = TypeVar("T")
+EE: TypeAlias = Union[Type[E], "ExpectedExceptionGroup[E]"]
+EEE: TypeAlias = Union[E, Type[E], "ExpectedExceptionGroup[E]"]
+
+
+# copy-pasting code between here and WIP pytest PR, which doesn't use future annotations
+# ruff: noqa: UP007
+
+
+# inherit from BaseExceptionGroup for the sake of typing the return type of raises.
+# Maybe it also works to do
+# `if TYPE_CHECKING: ExpectedExceptionGroup = BaseExceptionGroup`
+# though then we would probably need to support `raises(ExceptionGroup(...))`
+# class ExpectedExceptionGroup(BaseExceptionGroup[E]):
+@final
+class ExpectedExceptionGroup(Generic[E]):
+    # one might consider also accepting `ExpectedExceptionGroup(SyntaxError, ValueError)`
+    @overload
+    def __init__(self, exceptions: EEE[E], *args: EEE[E]):
+        ...
+
+    @overload
+    def __init__(self, exceptions: tuple[EEE[E], ...]):
+        ...
+
+    def __init__(self, exceptions: EEE[E] | tuple[EEE[E], ...], *args: EEE[E]):
+        if isinstance(exceptions, tuple):
+            if args:
+                raise ValueError(
+                    "All arguments must be exceptions if passing multiple positional arguments."
+                )
+            self.expected_exceptions = exceptions
+        else:
+            self.expected_exceptions = (exceptions, *args)
+        if not all(
+            isinstance(exc, (type, BaseException, ExpectedExceptionGroup))
+            for exc in self.expected_exceptions
+        ):
+            raise ValueError(
+                "All arguments must be exception instances, types, or ExpectedExceptionGroup."
+            )
+
+    # TODO: TypeGuard
+    def matches(
+        self,
+        exc_val: Optional[BaseException],
+    ) -> bool:
+        if exc_val is None:
+            return False
+        if not isinstance(exc_val, BaseExceptionGroup):
+            return False
+        if len(exc_val.exceptions) != len(self.expected_exceptions):
+            return False
+        remaining_exceptions = list(self.expected_exceptions)
+        for e in exc_val.exceptions:
+            for rem_e in remaining_exceptions:
+                # TODO: how to print string diff on mismatch?
+                # Probably accumulate them, and then if fail, print them
+                if (
+                    (isinstance(rem_e, type) and isinstance(e, rem_e))
+                    or (
+                        isinstance(e, BaseExceptionGroup)
+                        and isinstance(rem_e, ExpectedExceptionGroup)
+                        and rem_e.matches(e)
+                    )
+                    or (
+                        isinstance(rem_e, BaseException)
+                        and isinstance(e, type(rem_e))
+                        and re.search(str(rem_e), str(e))
+                    )
+                ):
+                    remaining_exceptions.remove(rem_e)  # type: ignore # ??
```

Comment:
What's the typechecker error?

Reply:
`error: Argument 1 to "remove" of "list" has incompatible type "Union[BaseException, Type[E], ExpectedExceptionGroup[E]]"; expected "Union[E, Type[E], ExpectedExceptionGroup[E]]"  [arg-type]`
i.e. `BaseException != E` where `E = TypeVar(BaseException, bound=BaseException)`

Is that something to do with the variance? (I still don't understand variance at all >.>)

Reply:
Yeah this has to do with variance. `list[int]` will not accept `.remove(<float>)`. Not sure how to fix it though.